doc/answer.1
.if n \{\
.	ds ct "
.\}
.if t \{\
.	ds ct \\(co
.\}
.TH ANSWER 1L "Elm Version 2.5" "USENET Community Trust"
.SH NAME
answer - phone message transcription system
.SH SYNOPSIS
.B answer
[
.B \-pu
]
.PP
.SH DESCRIPTION
.I Answer\^
is a part of the 
.B Elm
mail system and is designed for secretaries and the like
to allow them to answer phones and|or listen to phone message
machines and quickly and easily transcribe the messages into
electronic mail.
.PP
The program will use standard mail addresses or aliases
included in the standard .elm directory alias files.
It can also use
its own alias set (usually a subset of the
system aliases, with more fields per user (i.e. unique first
names, unique last names, first-initial last name, and so on)).
Note that if you want to use this capability, it
means that you must create this alias set and then install it, using
.I newalias,
as a separate alias table for the account this program is to 
be run from.
.PP
Options are:
.TP 1.0i
.B "  -p"
Prompt for phone slip type message fields.
.TP
.B "  -u"
Allow names that are not in the alias table.  Without this
option all names must be resolvable as aliases and answer will
reprompt for a new name if the alias cannot be resolved.
.PP
Interface-wise, the program is a breeze to use - type 'answer'
to start it up and then merely type a user name at the
.nf
	
	Message to:

.fi
prompt, fill in the requested information (simply type an
"x" next to the pertinent information - "TELEPHONED",
"PLEASE CALL", "URGENT", etc. if the -p option was given), 
and then enter a message, if any, ending with a blank line.
If you elect to use the special alias file, 
names can be entered as 'Firstname Lastname' and the system
will try to find a mail alias based on the first letter of the
first name plus the last name.  For example, 'Dave Taylor'
will cause the program to search for an alias 'd_taylor' in
the file (note the transliteration into lower case).   The 
program will loop until killed or quit.
.PP
.I Answer
can be easily exited by typing any of 'quit', 'exit' 'done', 
or 'bye' at the 'Message to:' prompt.
.SH AUTHOR
Elm Development Group
.SH SEE ALSO
.I "The Elm Alias System Users Guide"
.sp
.br
newalias(1L), checkalias(1L), mail(1), mailx(1), frm(1L), printmail(1L)
.SH FILES
$HOME/.elm/aliases.text       individual alias source
.br
$HOME/.elm/aliases.dir        individual alias dbz directory table
.br
$HOME/.elm/aliases.pag        individual alias dbz hash table
.br
$HOME/.elm/aliases            individual alias data table
.br
/tmp/snd.$$                   edit buffer for outgoing mail
.SH BUG REPORTS TO
Bill Pemberton  flash@virginia.edu
.SH COPYRIGHTS
\fB\*(ct\fRCopyright 1988-1995 by The USENET Community Trust
.br
Derived from Elm 2.0, \fB\*(ct\fR Copyright 1986, 1987 by Dave Taylor


Changes
Changes in 2.5 PL8

- fixed a hang when hitting certain 8-bit chars in the From line

- fixed a file descriptor leak (from Andreas Luik)

- removed all Expires: processing


Changes in 2.5 PL7

- a few display issues fixed (from Kenneth Reek)

- buffer overflow in frm fixed

Changes in 2.5 PL6

- Fixed message-id in the In-Reply-To header.

- Actually show the -r option in the usage

- Whacked a few more places where we might follow a symlink in tmpdir

- Fastmail didn't grok commas correctly.

- Change lock.c to not use errno unless there's an errno

Changes in 2.5 PL5

- A few more buffer overruns fixed.

Changes in 2.5 PL4

- little bug in aliasdb.c (from Scot W. Hetzel <scot@genroco.com>)

- Message-Id: could overflow (from Darin Wayrynen <darin@deru.net>)

- Fixed the permissions setting problem with the edit mailbox function

- Added a ResizeScreen() call after show_msg() calls to hopefully deal wit
  SIGWINCH problems (inspired from Yuke Zhuge <zhuge@stone.pdx.informix.com>)
  
- Fixed problems under SunOS 4.1.3 (from Paul J. Menchini <mench@mench.com>)

- Fixed content encoding (from David Woolley  <david@djwhome.demon.co.uk>)

- Attempt to stop crashes with long recipient list (from Gilles Detillieux 
  <grdetil@scrc.umanitoba.ca>)
  

Changes in 2.5 PL3

- Fixed the declaration of outbound_from_name (from Richard Jackson 
  <rjackson@gmu.edu>)

- Fixed the leapyear macro (from Gerhard Niklasch)
  
- Fixed editing of a long CC line (from Derek Godfrey <djg@sjc.ascend.com>)

- Mail that lands in the sentmail with attachments now has the 
  Content-disposition header text added to the message (from Dave 
  Spensley)
  
- Made allto a bit longer (suggested by Dave Spensley 
  <davesp@typhoon.xnet.com>)

- Fixed Configure to find (again) fcntl on Linux

- can_access will now allow you to write to a char device (/dev/null)

Changes in 2.5 PL2

- Fixed (again, maybe?) savebyalias from Adam Benjamin <aebenjam@opentext.com>

- Fixed bug were an alias delete would crash elm if the mailbox was empty

Changes in 2.5 PL1

- Fixed savebyalias from Michael D Hughes <mdh@logcabin.woods.bridge.com> 
  and Robert Schwartz <rjs@elt.com>
  
- Changes for 8 bit clean from H.J. Lu <hjl@varesearch.com>

Changes in 2.5 PL0

- One minor change to make an easier compile on SparcLinux.

Changes in 2.5 PL0pre8

- Y2K stuff from Ed Symanzik <zik@msu.edu>

Changes in 2.5 PL0pre7

- minor change to deal with egcs

Changes in 2.5 PL0pre6

- newalias nows does a geteuid() instead of getuid().  This is to lets 
  SysAdmins do "su <user> -c 'newalias'"
  

Changes in 2.5 PL0pre5

- Fix for GECOS entries with comma-delimited sub-fields from Chip 
  Rosenthal <chip@unicom.com>.

- Fix? for keys getting screw up after a ^Z from Chip Rosenthal
  <chip@unicom.com>.


Changes in 2.5 PL0pre4

- A few changes for BSD from Chip Rosenthal <chip@unicom.com>.

- A few changes to Configure suggested by Chip Rosenthal <chip@unicom.com>.


Changes in 2.5 PL0pre3

- Configure didn't see fcntl locking on Linux.  Fix from Sean McDermott 
  <spm@intel.com>


Changes in 2.5 beta 3

- Fixed a problem where elm would try to close a file twice when
  sending an attachment (from reichler@ews.uiuc.edu)

- Added docs for environment variables to the man page (from Eric
  Haszlakiewicz <nimenees@uaccess.net>)

Changes in 2.5 beta 2

- IncomingFolders no longer set Magic Mode on for everything

- Configure should now deal with Linux.

- Make //pattern honor current limit (from Keith Neufeld<neufeld@pvi.org>)

- year 2000 fix.....?

- Building on Solaris requires /usr/ccs/bin to be in the PATH

Changes in 2.5 beta 1

- b)ounce honor DONT_ADD_FROM (from Keith Neufeld<neufeld@pvi.org>)

- removed [encode] display and reply (from Keith Neufeld<neufeld@pvi.org>)

- make right/left arrows work in the file browser the same as in the
  message index. (from Keith Neufeld<neufeld@pvi.org>)

- fixed a bug in the incomingfolders code (it wouldn't save) (from Bob Schwartz
  <rjs@elt.com>)
  
- fixed the meaning of noheader for forwards (from Keith Neufeld<neufeld@pvi.org>)



Changes in 2.5 alpha 11

- fixed a bug in the alias code (thanks to Keith Neufeld<neufeld@pvi.org>)

- fixed the generation of temp file names in newmbox.c

- allowed longer Cc: lines.  This was due to a crash with a group reply.

- added incomingfolders to the elmrc.  This is a list of folders that will
  automatically be opened with "magic mode" on. (inspired by a patch from
  Cristophe Kalt <kalt@ensta.fr>)


Changes in 2.5 alpha 10

- Press u on a message that is read but not deleted will mark the message as
  unread (from Larry Schwimmer <rosebud@cyclone.Stanford.EDU>).

- mmencode.1 man page is now installed with others (from Gregory Margo
  <gmargo@newton.vip.best.com>).

- Magic mode stuff is now in help and NLS (from Gregory Margo
  <gmargo@newton.vip.best.com>).

- allow elimination of aliases from larger aliases (To: bigalias -subalias).
  Based on incomplete patch by Waldi Ravens <waldi@moacs.indiv.nl.net>.  Also
  removes redundant declarations of get_alias_address(), which is declared in
  elm_globals.h (from Keith Neufeld<neufeld@pvi.org>)

- On an entry field that normally clears to be retyped from scratch as soon
  as the user presses a key (Subject: on header edit screen), allow right
  arrow and delete to begin editing the existing default answer. (from Keith
  Neufeld<neufeld@pvi.org>)

- Add a warning to Configure about disabling shell escapes (from Keith
  Neufeld<neufeld@pvi.org>)

- Eliminate complaints in doc/Makefile about duplicate makefile entries if
  mansrc and catmansrc are both "none" (from Keith Neufeld<neufeld@pvi.org>)

- Allow entry of uppercase commands on header display/edit screens, since
  that's how they're parenthesized on the display.  (from Bob Schwartz
  <rjs@elt.com> and Keith Neufeld <neufeld@pvi.org>

- Make 2+ -line error messages write over more of the screen instead of

utils/answer.c


/*******************************************************************************
 *  The Elm Mail System  -  $Revision: 1.5 $   $State: Exp $
 *
 *                      Copyright (c) 1988-1995 USENET Community Trust
 * 			Copyright (c) 1986,1987 Dave Taylor
 *******************************************************************************
 * Bug reports, patches, comments, suggestions should be sent to:
 *
 *      Bill Pemberton, Elm Coordinator
 *      flash@virginia.edu
 *
 *******************************************************************************
 * $Log: answer.c,v $
 * Revision 1.5  1999/03/24  14:04:12  wfp5p
 * elm 2.5PL0
 *
 * Revision 1.4  1996/03/14  17:30:06  wfp5p
 * Alpha 9
 *
 * Revision 1.3  1995/09/29  17:42:41  wfp5p
 * Alpha 8 (Chip's big changes)
 *
 * Revision 1.2  1995/04/20  21:02:04  wfp5p
 * Added the showreply feature and emacs key bindings.
 *
 * Revision 1.1.1.1  1995/04/19  20:38:40  wfp5p
 * Initial import of elm 2.4 PL0 as base for elm 2.5.
 *
 ******************************************************************************/

/** This program is a phone message transcription system, and
    is designed for secretaries and the like, to allow them to
    painlessly generate electronic mail instead of paper forms.

    Note: this program ONLY uses the local alias file, and does not
	  even read in the system alias file at all.

**/

#define INTERN
#include "elm_defs.h"
#include "s_answer.h"
#include "ndbz.h"

#define  ELM		"elm"		/* where the elm program lives */

int user_data;		/* fileno of user data file   */
DBZ *hash;		/* dbz file for same */

char *get_alias_address(), *get_token(), *strip_parens(), *shift_lower();

static char *quit_word, *exit_word, *done_word, *bye_word;

main(argc, argv)
int argc;
char *argv[];
{
	FILE *fd;
	char *address, buffer[LONG_STRING], tempfile[SLEN], *cp;
	char  name[SLEN], recip_name[SLEN], in_line[SLEN];
	int   msgnum = 0, eof, allow_name = 0, phone_slip = 0;
	int   ans_pid = getpid();

	initialize_common();
	
	quit_word = catgets(elm_msg_cat, AnswerSet, AnswerQuitWord, "quit");
	exit_word = catgets(elm_msg_cat, AnswerSet, AnswerExitWord, "exit");
	done_word = catgets(elm_msg_cat, AnswerSet, AnswerDoneWord, "done");
	bye_word = catgets(elm_msg_cat, AnswerSet, AnswerByeWord, "bye");
/*
 *	simplistic crack arguments, looking for -u/-p
 *	-u = allow user names not in alias table
 *	-p = prompt for phone slip messages
 */
	for (msgnum = 1; msgnum < argc; msgnum++) {
	  if (istrcmp(argv[msgnum], "-u") == 0)
	    allow_name = 1;
	  if (istrcmp(argv[msgnum], "-p") == 0)
	    phone_slip = 1;
	  if (istrcmp(argv[msgnum], "-pu") == 0) {
	    allow_name = 1;
	    phone_slip = 1;
	  }
	  if (istrcmp(argv[msgnum], "-up") == 0) {
	    allow_name = 1;
	    phone_slip = 1;
	  }
	}

	open_alias_file();

	while (1) {
	  if (msgnum > 9999) msgnum = 0;
	
	  printf("\n-------------------------------------------------------------------------------\n");

prompt:   printf(catgets(elm_msg_cat, AnswerSet, AnswerMessageTo, "\nMessage to: "));
	  if (fgets(recip_name, SLEN, stdin) == NULL) {
		putchar('\n');
		exit(0);
	  }
	  if(recip_name[0] == '\0')
	    goto prompt;
	  
	  cp = &recip_name[strlen(recip_name)-1];
	  if(*cp == '\n') *cp = '\0';
	  if(recip_name[0] == '\0')
		goto prompt;

	  if ((istrcmp(recip_name, quit_word) == 0) ||
	      (istrcmp(recip_name, exit_word) == 0) ||
	      (istrcmp(recip_name, done_word) == 0) ||
	      (istrcmp(recip_name, bye_word)  == 0))
	     exit(0);

	  if (translate(recip_name, name) == 0)
	    goto prompt;

	  address = get_alias_address(name, 1, 0);

	  if (address == NULL || strlen(address) == 0) {
	    if (allow_name)
	      address =  name;
	    else {
	      printf(catgets(elm_msg_cat, AnswerSet, AnswerSorryNotFound,
		     "Sorry, could not find '%s' [%s] in list!\n"),
		     recip_name, name);
	      goto prompt;
	    }
	  }

	  printf("address '%s'\n", address);

	  sprintf(tempfile, "%sans.%d.%d", default_temp, ans_pid, msgnum++);

	  if ((fd = fopen(tempfile,"w")) == NULL)
	    exit(printf(catgets(elm_msg_cat, AnswerSet, AnswerCouldNotOpenWrite,
			"** Fatal Error: could not open %s to write\n"),
		 tempfile));


	/** Enter standard phone message fields **/
	  if (phone_slip) {
	    strcpy(buffer, catgets(elm_msg_cat, AnswerSet, AnswerCaller, "Caller: "));
	    printf("\n%s",buffer);
	    fflush(stdout);
	    fgets(in_line, SLEN, stdin);
	    if (strlen(in_line) > 1)
	      fprintf(fd,"%s%s",buffer,in_line);

	    strcpy(buffer, catgets(elm_msg_cat, AnswerSet, AnswerOf, "of:     "));
	    printf("%s",buffer);
	    fflush(stdout);
	    fgets(in_line, SLEN, stdin);
	    if (strlen(in_line) > 1)
	      fprintf(fd,"%s%s",buffer,in_line);

	    strcpy(buffer, catgets(elm_msg_cat, AnswerSet, AnswerPhone, "Phone:  "));
	    printf("%s",buffer);
	    fflush(stdout);
	    fgets(in_line, SLEN, stdin);
	    if (strlen(in_line) > 1)
	      fprintf(fd,"%s%s\n",buffer,in_line);

	    strcpy(buffer, catgets(elm_msg_cat, AnswerSet, AnswerTelephoned, "TELEPHONED         - "));
	    printf("\n%s",buffer);
	    fflush(stdout);
	    fgets(in_line, SLEN, stdin);
	    if (strlen(in_line) > 1)
	      fprintf(fd,"%s%s",buffer,in_line);

	    strcpy(buffer, catgets(elm_msg_cat, AnswerSet, AnswerCalledToSeeYou, "CALLED TO SEE YOU  - "));
	    printf("%s",buffer);
	    fflush(stdout);
	    fgets(in_line, SLEN, stdin);
	    if (strlen(in_line) > 1)
	      fprintf(fd,"%s%s",buffer,in_line);

	    strcpy(buffer, catgets(elm_msg_cat, AnswerSet, AnswerWantsToSeeYou, "WANTS TO SEE YOU   - "));
	    printf("%s",buffer);
	    fflush(stdout);
	    fgets(in_line, SLEN, stdin);
	    if (strlen(in_line) > 1)
	      fprintf(fd,"%s%s",buffer,in_line);

	    strcpy(buffer, catgets(elm_msg_cat, AnswerSet, AnswerReturnedYourCall, "RETURNED YOUR CALL - "));
	    printf("%s",buffer);
	    fflush(stdout);
	    fgets(in_line, SLEN, stdin);
	    if (strlen(in_line) > 1)
	      fprintf(fd,"%s%s",buffer,in_line);

	    strcpy(buffer, catgets(elm_msg_cat, AnswerSet, AnswerPleaseCall, "PLEASE CALL        - "));
	    printf("%s",buffer);
	    fflush(stdout);
	    fgets(in_line, SLEN, stdin);
	    if (strlen(in_line) > 1)
	      fprintf(fd,"%s%s",buffer,in_line);

	    strcpy(buffer, catgets(elm_msg_cat, AnswerSet, AnswerWillCallAgain, "WILL CALL AGAIN    - "));
	    printf("%s",buffer);
	    fflush(stdout);
	    fgets(in_line, SLEN, stdin);
	    if (strlen(in_line) > 1)
	      fprintf(fd,"%s%s",buffer,in_line);

	    strcpy(buffer, catgets(elm_msg_cat, AnswerSet, AnswerUrgent, "*****URGENT******  - "));
	    printf("%s",buffer);
	    fflush(stdout);
	    fgets(in_line, SLEN, stdin);
	    if (strlen(in_line) > 1)
	      fprintf(fd,"%s%s",buffer,in_line);
	  }

	  printf(catgets(elm_msg_cat, AnswerSet, AnswerEnterMessage,
		"\n\nEnter message for %s ending with a blank line.\n\n"), 
		 recip_name);

	  fprintf(fd,"\n\n");

	  do {
	   printf("> ");
	   if (! (eof = (fgets(buffer, SLEN, stdin) == NULL))) 
	     fprintf(fd, "%s", buffer);
	  } while (! eof && strlen(buffer) > 1);
	
	  fclose(fd);
 
	  sprintf(buffer, catgets(elm_msg_cat, AnswerSet, AnswerElmCommand,
	     "( ( %s -s \"While You Were Out\" %s < %s ; %s %s) & ) > /dev/null"),
	     ELM, strip_parens(address), tempfile, remove_cmd, tempfile);

	  system(buffer);
	}
}

int
translate(fullname, name)
char *fullname, *name;
{
	/** translate fullname into name..
	       'first last'  translated to first_initial - underline - last
	       'initial last' translated to initial - underline - last
	    Return 0 if error.
	**/
	register int i, lastname = 0, len;

	for (i=0, len = strlen(fullname); i < len; i++) {

	  fullname[i] = tolower(fullname[i]);

	  if (fullname[i] == ' ') 
	    if (lastname) {
	      printf(catgets(elm_msg_cat, AnswerSet, AnswerCannotHaveMoreNames,
	      "** Can't have more than 'FirstName LastName' as address!\n"));
	      return(0);
	    }
	    else
	      lastname = i+1;
	
	}

	if (lastname) 
	  sprintf(name, "%c_%s", fullname[0], (char *) fullname + lastname);
	else
	  strcpy(name, fullname);

	return(1);
}

	    
open_alias_file()
{
	/** open the user alias file **/

	char fname[SLEN];

	sprintf(fname,  "%s/.elm/aliases", getenv("HOME")); 

	if ((hash = dbz_open(fname, O_RDONLY, 0)) == NULL) 
	  exit(printf("** Fatal Error: Could not open %s!\n", fname));

	if ((user_data = open(fname, O_RDONLY)) == -1) 
	  return;
}

char *get_alias_address(name, mailing, depth)
char *name;
int   mailing, depth;
{
	/** return the line from either datafile that corresponds 
	    to the specified name.  If 'mailing' specified, then
	    fully expand group names.  Returns NULL if not found.
	    Depth is the nesting depth, and varies according to the
	    nesting level of the routine.  **/

	static char buffer[VERY_LONG_STRING];
	static char sprbuffer[VERY_LONG_STRING];
	datum  key, value;
	int    loc;
	struct alias_disk_rec entry;

	name = shift_lower(name);
	key.dptr = name;
	key.dsize = strlen(name);
	value = dbz_fetch(hash, key);
	if (value.dptr == NULL)
	    return( (char *) NULL); /* not found */

	bcopy(value.dptr, (char *) &loc, sizeof(loc));
	loc -= sizeof(entry);
	lseek(user_data, loc, SEEK_SET);
	read(user_data, (char *) &entry, sizeof(entry));
	read(user_data, buffer, entry.length > VERY_LONG_STRING ? VERY_LONG_STRING : entry.length);
	if ((entry.type & GROUP) != 0 && mailing) {
	    if (expand_group(sprbuffer, buffer + entry.address,
			     depth) < 0)
		return NULL;
	} else {
	    sprintf(sprbuffer, "%s (%s)", buffer + entry.address,
		    buffer + entry.name);
	}
	return sprbuffer;
}

int expand_group(target, members, depth)
char *target;
char *members;
int   depth;
{
	/** given a group of names separated by commas, this routine
	    will return a string that is the full addresses of each
	    member separated by spaces.  Depth is the current recursion
	    depth of the expansion (for the 'get_token' routine) **/

	char   buf[VERY_LONG_STRING], *word, *address, *bufptr;

	strcpy(buf, members); 	/* parameter safety! */
	target[0] = '\0';	/* nothing in yet!   */
	bufptr = (char *) buf;	/* grab the address  */
	depth++;		/* one more deeply into stack */

	while ((word = (char *) get_token(bufptr, "!, ", depth)) != NULL) {
	  if ((address = (char *) get_alias_address(word, 1, depth)) == NULL) {
	    fprintf(stderr, catgets(elm_msg_cat, AnswerSet, AnswerNotFoundForGroup,
		"Alias %s not found for group expansion!\n"), word);
	    return -1;
	  }
	  else if (strcmp(target,address) != 0) {
	    sprintf(target + strlen(target), " %s", address);
	  }

	  bufptr = NULL;
	}
	return 0;
}

print_long(buffer, init_len)
char *buffer;
int   init_len;
{
	/** print buffer out, 80 characters (or less) per line, for
	    as many lines as needed.  If 'init_len' is specified, 
	    it is the length that the first line can be.
	**/

	register int i, loc=0, space, length, len; 

	/* In general, go to 80 characters beyond current character
	   being processed, and then work backwards until space found! */

	length = init_len;

	do {
	  if (strlen(buffer) > loc + length) {
	    space = loc + length;
	    while (buffer[space] != ' ' && space > loc + 50) space--;
	    for (i=loc;i <= space;i++)
	      putchar(buffer[i]);
	    putchar('\n');
	    loc = space;
	  }
	  else {
	    for (i=loc, len = strlen(buffer);i < len;i++)
	      putchar(buffer[i]);
	    putchar('\n');
	    loc = len;
	  }
	  length = 80;
	} while (loc < strlen(buffer));
}

/****
     The following is a newly chopped version of the 'strtok' routine
  that can work in a recursive way (up to 20 levels of recursion) by
  changing the character buffer to an array of character buffers....
****/

#define MAX_RECURSION		20		/* up to 20 deep recursion */

#undef  NULL
#define NULL			(char *) 0	/* for this routine only   */


char *get_token(string, sepset, depth)
char *string, *sepset;
int  depth;
{

	/** string is the string pointer to break up, sepstr are the
	    list of characters that can break the line up and depth
	    is the current nesting/recursion depth of the call **/

	register char	*p, *q, *r;
	static char	*savept[MAX_RECURSION];

	/** is there space on the recursion stack? **/

	if (depth >= MAX_RECURSION) {
	 fprintf(stderr, catgets(elm_msg_cat, AnswerSet, AnswerRecursionTooDeep,
		"Error: Get_token calls nested greater than %d deep!\n"),
			MAX_RECURSION);
	 exit(1);
	}

	/* set up the pointer for the first or subsequent call */
	p = (string == NULL)? savept[depth]: string;

	if(p == 0)		/* return if no tokens remaining */
		return(NULL);

	q = p + strspn(p, sepset);	/* skip leading separators */

	if (*q == '\0')		/* return if no tokens remaining */
		return(NULL);

	if ((r = strpbrk(q, sepset)) == NULL)	/* move past token */
		savept[depth] = 0;	/* indicate this is last token */
	else {
		*r = '\0';
		savept[depth] = ++r;
	}
	return(q);
}
